Minor: Handle special error format for create_upload endpoint. #22

ylecuyer opened the pull request:
Fix for #22 

It adds a fallback for special error format in create_upload endpoint.

ylecuyer commented on lib/strava/errors/fault.rb:
```diff
@@ -2,15 +2,17 @@ module Strava
-        response[:body]['errors']
+        return response[:body]['errors'] if response[:body]['errors']
+
+        [response[:body]['error']] if response[:body]['error']
```
I wrap the error in an array so that it stays consistent with normal error format

dblock commented:
This feels incorrect. The API has predictable response and we're trying to be too smart about handling it. 

According to Strava documentation this endpoint returns a [Fault](https://developers.strava.com/docs/reference/#api-models-Fault) which returns only `errors` and `message`. Are you getting a 400 back here with something else? 

Do you have a VCR / HTTP request/response from Strava for this?

ylecuyer commented:
Here is an example with a missing field which is using the standard Fault format: 

```
ylecuyer@inwin:~/Projects/OnMove200$ http --form POST "https://www.strava.com/api/v3/uploads" file@/home/ylecuyer/Desktop/DATA/strava.gpx name='test for #22' external_id='#22' "Authorization: Bearer REDACTED"
HTTP/1.1 400 Bad Request
Cache-Control: no-cache
Connection: keep-alive
Content-Type: application/json; charset=utf-8
Date: Fri, 20 Dec 2019 22:46:05 GMT
Referrer-Policy: strict-origin-when-cross-origin
Status: 400 Bad Request
Transfer-Encoding: chunked
Vary: Origin
Via: 1.1 linkerd
X-Content-Type-Options: nosniff
X-Download-Options: noopen
X-Frame-Options: SAMEORIGIN,DENY
X-Permitted-Cross-Domain-Policies: none
X-RateLimit-Limit: 600,30000
X-RateLimit-Usage: 2,3
X-Request-Id: edbe7eab-9170-4a1f-8f9e-b096bc3286ad
X-XSS-Protection: 1; mode=block
content-encoding: gzip

{
    "errors": [
        {
            "code": "required",
            "field": "data_type",
            "resource": "Upload"
        }
    ],
    "message": "Bad Request"
}
```

Here is the duplicate activity error which doesn't follow the fault format

```
ylecuyer@inwin:~/Projects/OnMove200$ http --form POST "https://www.strava.com/api/v3/uploads" file@/home/ylecuyer/Desktop/DATA/strava.gpx name='test for #22' data_type='gpx' external_id='#22' "Authorization: Bearer REDACTED"
HTTP/1.1 400 Bad Request
Cache-Control: no-cache
Connection: keep-alive
Content-Type: application/json; charset=utf-8
Date: Fri, 20 Dec 2019 22:45:29 GMT
Referrer-Policy: strict-origin-when-cross-origin
Status: 400 Bad Request
Transfer-Encoding: chunked
Vary: Origin
Via: 1.1 linkerd
X-Content-Type-Options: nosniff
X-Download-Options: noopen
X-Frame-Options: SAMEORIGIN,DENY
X-Permitted-Cross-Domain-Policies: none
X-RateLimit-Limit: 600,30000
X-RateLimit-Usage: 1,2
X-Request-Id: e48ccecd-5dae-4b9a-8409-1967187d38a7
X-XSS-Protection: 1; mode=block
content-encoding: gzip

{
    "activity_id": null,
    "error": "#22.gpx duplicate of activity 2941971481",
    "external_id": "#22.gpx",
    "id": 3136495967,
    "id_str": "3136495967",
    "status": "There was an error processing your activity."
}
```

I wont take the time to deal with strava support, feel free to contact them if you want to be sure this is the intended behavior on their end.

dblock commented:
I've opened a ticket with Strava.

dblock commented:
<img width="1057" alt="Screen Shot 2019-12-24 at 11 14 10 AM" src="https://user-images.githubusercontent.com/542335/71419655-8803ce00-263e-11ea-9917-3618c95ffebc.png">

dblock commented:
Given what Strava said we should fix this here, but not by modifying the base class. It's a special case for `upload`, so we can handle it there.

I tried to write a spec for this, adding to create_upload_spec.rb:

```ruby
  let(:file) { 'spec/fixtures/strava/files/17611540601.tcx' }
  it 'handles duplicate activities', vcr: { cassette_name: 'client/create_upload_dup' } do
    upload = client.create_upload(
      file: Faraday::UploadIO.new(file, 'application/tcx+xml'),
      name: 'Uploaded Activity',
      description: 'Uploaded by strava-ruby-client.',
      data_type: 'tcx',
      external_id: 'strava-ruby-client-upload-1'
    )
    ...
end
```

I have pre-uploaded `spec/fixtures/strava/files/17611540601.tcx`, but I cannot get a duplicate error back. I'm getting a 201 with the original activity.
 
Are you still able to reproduce this? Can you turn that into a spec?

ylecuyer commented:
> Did they change this behavior? Are you still able to reproduce this? Can you turn that into a spec?

They didn't change the behavior and can reproduce this error every time.

I tried to do the change with a new error only for the `/uploads` endpoint, I'm not fan of the end result but if the idea is ok for you I'll add some tests and ruboclean it.

dblock commented:
I like this OK, so would merge something like this. The URL being a structure, I'd make sure the API verb is POST and the path is `/uploads`, rather than just checking `.to_s` which seems brittle.

dblock commented:
Better. It needs a spec and a rubocop fix, see failed build.

dblock commented:
Let's also add a comment in this code linking to the issue/conversation/something.

ylecuyer commented:
> Better. It needs a spec and a rubocop fix, see failed build.

I have added the spec and fixed the build :)
I disabled some rubocop rules locally which are irrelevants for spec files.

> Let's also add a comment in this code linking to the issue/conversation/something.

IMO this is better handled by git blame and we will have the issue number in the changelog and commit name

dblock commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 * [#18](https://github.com/dblock/strava-ruby-client/pull/18): Testing against Ruby 2.5.3 and 2.6.0 - [@lucianosousa](https://github.com/lucianosousa).
 * [#18](https://github.com/dblock/strava-ruby-client/pull/18): Upgraded Rubocop to 0.61.1 - [@lucianosousa](https://github.com/lucianosousa).
 * [#21](https://github.com/dblock/strava-ruby-client/pull/21): Include headers in error response - [@jameschevalier](https://github.com/jameschevalier).
+* [#23](https://github.com/dblock/strava-ruby-client/pull/23): Minor: handle special error format for create_upload endpoint. #22 - [@ylecuyer](https://github.com/ylecuyer).
```
This is my no means minor. You’ve fixed a bug in which upload errors aren’t properly handled. So just say “Fixed handling of create_upload errors”.

I typically also do [#23](...), [#22](...) and no period in the sentence.

dblock commented on lib/strava/web/raise_error.rb:
```diff
@@ -16,6 +18,10 @@ def on_complete(env)
         end
 
+        def handle_upload_errors(env)
+          raise Strava::Errors::UploadFailed, response_values(env) if env.status != 201 && env.body.key?('error') && !env.body['error'].blank?
```
Turn this into a guard (ruby bear practice)

```
return unless/if ...

raise ...
```

Also easier to read.

dblock commented on spec/strava/api/client/endpoints/uploads/create_upload_spec.rb:
```diff
@@ -17,4 +17,18 @@
+        file: Faraday::UploadIO.new(file, 'application/tcx+xml'),
+        data_type: 'tcx'
+      )
+    end.to raise_error(Strava::Errors::UploadFailed) do |error| # rubocop:disable Style/MultilineBlockChain
```
For rubocop run `rubocop -a ; rubocop --auto-gen-config`, don’t suppress inline.

dblock commented on spec/strava/api/client/endpoints/uploads/create_upload_spec.rb:
```diff
@@ -17,4 +17,18 @@
     expect(upload.status).to eq 'Your activity is still being processed.'
     expect(upload.activity_id).to eq nil
   end
+  it 'returns a Strava::Errors::UploadFailed on upload error', vcr: :all do
```
What’s this vcr: all? Are you missing a VCR file for the error explicitly recorded in the commit? Make sure that doesn’t have a key, etc.

simonneutert commented:
@dblock could you move this pr in a separate branch, so I can try and fix your last requests, before this is ready to merge?

dblock commented:
> @dblock could you move this pr in a separate branch, so I can try and fix your last requests, before this is ready to merge?

You don't need me for this. Make a branch in your fork, `git pull` these changes, make any updates and open a new PR.

dblock commented:
@simonneutert Want to finish this one too?

simonneutert commented:
turns out Strava seemed to have changed to async "background jobs", so they accept the file if nothing is totally wrong. handing out an upload ID for further inquiries.

- when reuploading a file, then querying via the new ID, an error stating "... duplicate ..." is part of the body

I need to reread what I came up with, then I prepare a PR for us to decide on how to finish this.

simonneutert commented:
@dblock take a glimpse in my draft, please https://github.com/dblock/strava-ruby-client/pull/67

As a final polish, I want to have the behavior of UploadFailedError much closer to `Fault` we are using already, but I think moving the error raising in the model is the way to go here.

dblock commented:
LGTM, just needs a rebase

dblock commented:
Closed via https://github.com/dblock/strava-ruby-client/pull/67
